### DISCUSSION TITLE
Converting exr files to ktx ?

### DISCUSSION THREAD

**Discussion by vis-prime** [KhronosGroup/KTX-Software#696] [Q&A] [answered]
i'm performing a batch convertion of some textures and some normal and displacement maps are in exr format

any way to convert them too ?

`failed to create image from F:\Poly Haven\wooden_rough_planks\textures\wooden_rough_planks_nor_gl_4k.exr. Format of input file "F:\Poly Haven\wooden_rough_planks\textures\wooden_rough_planks_nor_gl_4k.exr" is unsupported.`

**Comment by MarkCallow** [comment]
Please wait a few more weeks. Some new tools are coming which will support EXR.

**Reply by arpu** [reply]
Hey i see there is  new alpha release, any news on this?

**Comment by MarkCallow** [comment, marked as answer]
The new `ktx create` tool in v4.3.0 supports creating .ktx2 files from EXR input.

**Reply by arpu** [reply]
Thanks!

**Reply by arpu** [reply]
Hey, is there any  samples how to use this? 

i tested with the EXR Version 1k from https://polyhaven.com/a/farm_sunset

with toktx and ktx create

``
toktx: failed to create image from /home/arpu/Downloads/farm_sunset_1k.exr(0,0).
Unsupported format 32-bit 0 needed.
``

``
 tools/ktx/ktx create /home/arpu/Downloads/farm_sunset_1k.exr test.ktx2  --format R8_UNORM
ktx create fatal: /home/arpu/Downloads/farm_sunset_1k.exr: Input file data type "exr_float" does not match the expected input data type of 8 bit "UNORM" for VK_FORMAT_R8_UNORM
``

master build adcccf152861158373311930e9b5e237d19168db

**Reply by donmccurdy** [reply]
@arpu these examples might show what you're looking for:

https://github.com/donmccurdy/KTX2-Samples/blob/98bfecb0a69891fc60c27f96a932f493c3a4406c/encode.sh#L27-L28

```bash
# float16
ktx create --format R16G16B16A16_SFLOAT --zstd 18 --assign-oetf linear --assign-primaries bt709 --generate-mipmap source/exr/2d_f16.exr ktx2/2d_rgba16_linear.ktx2

# float32
ktx create --format R32G32B32A32_SFLOAT --zstd 18 --assign-oetf linear --assign-primaries bt709 --generate-mipmap source/exr/2d_f32.exr ktx2/2d_rgba32_linear.ktx2
```

**Reply by arpu** [reply]
Thank you @donmccurdy 

what i see is i cannot encode exr files to  basis-z or uastc,  the idea is to use exr for threejs enviroment maps and background ( compressed -> smaller files for web)

**Reply by donmccurdy** [reply]
@arpu that won't be possible, no. EXR files are typically open domain [0, ∞] HDR data. BasisLZ and UASTC are limited to closed domain [0,1] and don't have sufficient bit depth for HDR, unless you do very custom/bespoke things to map your HDR data into RGBA8 ranges.

The commands above create float or half-float data compressed with Zstd, which is a more appropriate way to store HDR data. three.js cannot read those files yet, but I expect to be adding that support fairly soon. The files may or may not come out smaller than source EXR files, and mipmaps would add some size, but I'd expect file parsing speed on the web to be faster than with EXR.

BC6H is also worth considering as a compressed target, although I don't know of any tools capable of converting EXR data to BC6H-in-KTX2 data yet.

- EDIT: Work in progress – https://github.com/mrdoob/three.js/pull/26642

**Reply by MarkCallow** [reply]
EXR support was not added to the legacy `toktx` tool as you discovered, @arpu. And as @donmccurdy wrote 8-bit UNORM data does not have sufficient bit-depth for HDR.

I'd happily accept a PR to add BC6H encoding + RDO to libktx.
